lib/ansible/plugins/action/validate_argument_spec.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2018 Red Hat
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError, AnsibleModuleError
+from ansible.plugins.action import ActionBase
+from ansible.module_utils import basic
+from ansible.module_utils.six import iteritems, string_types
+
+
+class AnsibleArgSpecError(AnsibleModuleError):
+    def __init__(self, *args, **kwargs):
+        self.argument_errors = kwargs.pop('argument_errors', [])
+        super(AnsibleArgSpecError, self).__init__(*args, **kwargs)
+
+
+class ArgSpecValidatingAnsibleModule(basic.AnsibleModule):
+    '''AnsibleModule but with overridden _load_params so it doesn't read from stdin/ANSIBLE_ARGS'''
+
+    def __init__(self, *args, **kwargs):
+        self._params = kwargs.pop('params', {})
+        # remove meta fields that aren't valid AnsibleModule args
+        self.arg_spec_name = kwargs.pop('name', None)
+        self.arg_spec_description = kwargs.pop('description', None)
+
+        self.arg_validation_errors = []
+        super(ArgSpecValidatingAnsibleModule, self).__init__(*args, **kwargs)
+
+    # AnsibleModule._load_params sets self.params from a global, so neuter it here
+    # so our passed in 'params' kwarg is used instead
+    def _load_params(self):
+        self.params = self._params
+
+    def fail_json(self, *args, **kwargs):
+        msg = kwargs.pop('msg', 'Unknown arg spec validation error')
+
+        clean_msg = basic.remove_values(msg, self.no_log_values)
+
+        self.arg_validation_errors.append(clean_msg)
+
+    def check_for_errors(self):
+        if not self.arg_validation_errors:
+            return
+
+        msg = 'Validation of arguments failed:\n%s' % '\n'.join(self.arg_validation_errors)
+        raise AnsibleArgSpecError(msg, argument_errors=self.arg_validation_errors)
+
+
+class ActionModule(ActionBase):
+    ''' Validate an arg spec'''
+
+    TRANSFERS_FILES = False
+
+    # WARNING: modifies argument_spec
+    def build_args(self, argument_spec, task_vars):
+        args = {}
+        for key, attrs in iteritems(argument_spec):
+            if attrs is None:
+                argument_spec[key] = {'type': 'str'}
+
+            if key in task_vars:
+                if isinstance(task_vars[key], string_types):
+                    value = self._templar.do_template(task_vars[key])
+                    if value:
+                        args[key] = value
+                else:
+                    args[key] = task_vars[key]
+            elif attrs:
+                if 'aliases' in attrs:
+                    for item in attrs['aliases']:
+                        if item in task_vars:
+                            args[key] = self._templar.do_template(task_vars[key])
```

Comment:
Minor bug here.
```suggestion
                            args[key] = self._templar.do_template(task_vars[item])
```

Reply:
The bigger issue with aliases is the value doesn't seem to make its way to the task vars. I'm not sure exactly how to update that. `Validator.validate()` modifies the origin parameters, but those new values need to get passed in.

Reply:
Same issue with `default`.

Reply:
I removed this bit of code dealing with `aliases` and `defaults`. This would have allowed default value setting via the arg spec. And aliases are not considered supported for this feature as of yet.